### ISSUE TITLE
E-mail address encoding leads to several cache items

### ISSUE THREAD

**Issue by Mynyx** [contao/contao#1620]
**Affected version(s)**

Contao 4.9

**Description**

The e-mail address encoding leads to to several cache items of the same page. Every page that contains an e-mail address is saved in a new version when the cache time expires, even though the content of the page has not actually changed. This causes that additional write operations in the cache are performed and several items in the storage are created.

The reason is that the e-mail address encoding contains a random component. https://github.com/contao/contao/blob/b9c1002fb70588936db66415dbfa286fd11617f6/core-bundle/src/Resources/contao/library/Contao/StringUtil.php#L336

I wonder if this random selection between the two encoding options really leads to a security gain or the e-mail address encoding could be implemented differently to gain advantages in caching.

**Comment by ausi** [comment]
We could switch between the two variants for every iteration, e.g.:

```php
foreach ($arrCharacters as $index => $strCharacter)
{
	$strEncoded .= sprintf((($index % 2) ? '&#x%X;' : '&#%s;'), Utf8::ord($strCharacter));
}
```

**Comment by Toflar** [comment]
Two questions here: 

1. @leofeyer why is this assigned to 4.9? I'm sure it's the same in 4.4, isn't it?

2. What exactly is the issue here? Why would we end up having multiple cache entries? Can you elaborate on that @Mynyx? The cache key is not related to the content of a response at all. If I'm requesting `/foobar.html` and cache that for 24 hours, the cache key is built based on `/foobar.html` (and vary headers) but not the content. 24 hours later that cache item is expired but the key is still going to be the same. So the randomness here shouldn't be an issue at all.

**Comment by leofeyer** [comment]
@Mynyx wrote 

> **Affected version(s)**
> 
> Contao 4.9

Maybe the cache behaves differently in Contao 4.4?

**Comment by Mynyx** [comment]
> 2. What exactly is the issue here? Why would we end up having multiple cache entries? Can you elaborate on that @Mynyx? The cache key is not related to the content of a response at all. If I'm requesting `/foobar.html` and cache that for 24 hours, the cache key is built based on `/foobar.html` (and vary headers) but not the content. 24 hours later that cache item is expired but the key is still going to be the same. So the randomness here shouldn't be an issue at all.

The key for the cache entry with the response content is the content digest!?

https://github.com/Toflar/psr6-symfony-http-cache-store/blob/5364b76033d23801f5dd39e6ccc21fec3d4ccf96/src/Psr6Store.php#L214-L222

https://github.com/Toflar/psr6-symfony-http-cache-store/blob/5364b76033d23801f5dd39e6ccc21fec3d4ccf96/src/Psr6Store.php#L497

**Comment by Toflar** [comment]
Yeah but it shouldn't matter very much because the digest is used to improve storage. If two or more responses return the exact same content, the value can be shared amongst those.
So `/foobar.html` and `/foobar2.html` **might**  share the same content.
It's true that the randomness here will reduce chances for this, so we still might want to fix this, yes.

The true issue here, however, is that content digests never seem to expire.
I will check this.

**Comment by Toflar** [comment]
Implemented expiration of content digest items in https://github.com/Toflar/psr6-symfony-http-cache-store/pull/26. I pinged a few people working on caching at Sulu and eZ Publish, so we'll see how this goes.

Anyway, we should still apply @ausi's proposal here. There's not really anything to discuss imho, so the `up for discussion` label can be removed from my point of view.

**Comment by leofeyer** [comment]
I think that using a predictable algorithm will weaken the protection, therefore I have added the label.

**Comment by Mynyx** [comment]
If I would implement an e-mail address crawler with the ability to decode the entites, I would decode them in the whole document. Both types of entities are easy to detect with Regex.

Why should I invest the effort to manually detect them in another way, especially if Contao is just one CMS among many?

**Comment by leofeyer** [comment]
As discussed in Mumble on April, 9th, we want to implement @ausi's suggestion (see https://github.com/contao/contao/issues/1620#issuecomment-609485510) to get rid of the randomness and still mitigate against simple spam bots that cannot handle both entity variants.

**Comment by Toflar** [comment]
BTW: Content digest cache expiration has been released with version 2.3.0 of my cache implementation :)
